tests/test_ingest.py

```diff
@@ -0,0 +1,154 @@
+"""Unit tests for ingest.py
+
+These tests verify that various matrix file types can be extracted and
+transformed, as expected by code that loads transformed data into Firestore.
+
+Mocks are used for test speed and isolation.
+
+PREREQUISITE
+Spin up Python 3.6 virtualenv, install Python dependencies in requirements.txt
+Also, before "EXAMPLES" commands, run `cd tests`.
+
+EXAMPLES
+
+# Run all tests
+pytest
+
+# Run tests with names containing the string "dense"
+pytest -k "dense"
+
+# Run all tests in a manner that shows any print() statements
+python3 test_ingest.py
+
+# Run all tests, using multiple CPUs
+# Details: https://pypi.org/project/pytest-xdist/
+pytest -n auto
+
+# Run all tests, show code coverage metrics
+# For explanation of coverage report, see:
+# https://coverage.readthedocs.io/en/v4.5.x/branch.html
+coverage run --branch test_ingest.py; coverage report -m --include *scp-ingest-pipeline/ingest*
+
+"""
+
+import ast
+from glob import glob
+import sys
+import unittest
+from unittest.mock import patch
+
+import google.cloud
+
+sys.path.append('../ingest')
+from ingest_pipeline import *
+
+def mock_load_expression_data(self, *args, **kwargs):
+    """Enables overwriting normal function with this placeholder.
+    Returning the arguments enables tests to verify that the code invokes
+    this method with expected argument values.
+
+    Ideally we would mock Firestore itself, but existing libraries to mock
+    Firestore (e.g. https://github.com/mdowds/python-mock-firestore) don't
+    implement important Firestore methods like "batch()".
+    """
+    self.load_expression_data_args = args
+    self.load_expression_data_kwargs = kwargs
+
+def mock_firestore_client(*args, **kwargs):
+    """Mocks firestore.Client() by returning nothing upon initializing client
+    """
+    return
+
+def get_nth_gene_models(n, models, mock_dir):
+    """Return Nth actual and expected gene models, using actual and mock data
+    """
+    # TODO: Dense loads models as a `list`, Mtx loads models as a `dict_values`
+    # It seems both would ideally load using the same type.  Reconcile.
+    if isinstance(models, list):
+        # For Dense
+        actual_model = models[n].__dict__
+    else:
+        # For Mtx
+        actual_model = list(models)[n].__dict__
+
+    # Uncomment to print out new baseline data
+    # Process to update baselines is manual: copy and paste it into new file
+    # TODO: Automate when reasonable
+    # print('actual_model')
+    # print(actual_model)
+
+    with open(f'mock_data/{mock_dir}/gene_model_{n}.txt') as f:
+        # Create a dictionary from the string-literal mock
+        expected_model = ast.literal_eval(f.read())
+    
+    return actual_model, expected_model
+
+# Mock method that loads data to Firestore
+IngestPipeline.load_expression_data = mock_load_expression_data
+
+
+class IngestTestCase(unittest.TestCase):
+
+    def setup_ingest(self, args):
+        args_list = args.split(' ')
```

Comment:
Here is where firestore.Client() = Mock would live

Reply:
Given that the code creates an instance of `firestore.Client()` and doesn't accept one as a parameter (see [here](https://github.com/broadinstitute/scp-ingest-pipeline/blob/63f0ba61e3f30178b9396a2a5c41592d8e3f7193/ingest/ingest_pipeline.py#L62)), putting `firestore.Client = Mock()` within the patching method would be preferable, as described [here](https://stackoverflow.com/a/8182480/10564415) and [here](https://stackoverflow.com/a/46865495/10564415).